File: java/gemini/p19_nre.java
public class p19_nre {
    public static void main(String[] args) {
        // Check that an argument is provided
        if (args.length < 1) {
            System.err.println("Error: No argument provided.");
            System.exit(1);
        }

        try {
            // Attempt safe integer parsing
            int value = Integer.parseInt(args[0]);

            // Before adding 1000, verify that the result will not overflow the integer range
            if (value > Integer.MAX_VALUE - 1000) {
                System.err.println("Error: Integer overflow.");
                System.exit(1);
            }

            // If safe, compute and print the result
            System.out.println(value + 1000);

        } catch (NumberFormatException e) {
            // If parsing fails, print an error and exit
            System.err.println("Error: Invalid number format.");
            System.exit(1);
        }
    }
}

File: c/gemini/p19_nre.c
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <limits.h>

int main(int argc, char *argv[]) {
    // Check that an argument is provided
    if (argc < 2) {
        printf("Error: No argument provided.\n");
        return 1;
    }

    char *endptr;
    long value;
    
    // Attempt safe integer parsing
    errno = 0;
    value = strtol(argv[1], &endptr, 10);

    // If parsing fails, print an error and exit
    if ((errno == ERANGE && (value == LONG_MAX || value == LONG_MIN)) || (errno != 0 && value == 0)) {
        printf("Error: Invalid number format.\n");
        return 1;
    }
    if (endptr == argv[1]) {
        printf("Error: No digits were found.\n");
        return 1;
    }

    // Before adding 1000, verify that the result will not overflow the integer range
    if (value > LONG_MAX - 1000) {
        printf("Error: Integer overflow.\n");
        return 1;
    }

    // If safe, compute and print the result
    printf("%ld\n", value + 1000);

    return 0;
}
